ethsign.go

```diff
@@ -296,19 +287,141 @@ func main() {
 				} else {
 					tx = types.NewTransaction(nonce, to, value, gasLimit, gasPrice, data)
 				}
-				
+
 				signed, err := wallet.SignTxWithPassphrase(*acct, passphrase, tx, chainID)
 				if err != nil {
 					return cli.NewExitError("ethsign: failed to sign tx", 1)
 				}
 
 				encoded, _ := rlp.EncodeToBytes(signed)
 				fmt.Println(hexutil.Encode(encoded[:]))
-				
+
+				return nil
+			},
+		},
+
+		cli.Command{
+			Name:    "message",
+			Aliases: []string{"msg"},
+			Usage:   "sign arbitrary data with header prefix",
+			Flags: []cli.Flag{
+				cli.StringSliceFlag{
+					Name:   "key-store",
+					Usage:  "path to key store",
+					EnvVar: "ETH_KEYSTORE",
+				},
+				cli.StringFlag{
+					Name:   "from",
+					Usage:  "address of signing account",
+					EnvVar: "ETH_FROM",
+				},
+				cli.StringFlag{
+					Name:  "passphrase-file",
+					Usage: "path to file containing account passphrase",
+				},
+				cli.StringFlag{
+					Name:  "data",
+					Usage: "hex data to sign",
+				},
+			},
+			Action: func(c *cli.Context) error {
+				requireds := []string{
+					"from", "data",
+				}
+
+				for _, required := range requireds {
+					if c.String(required) == "" {
+						return cli.NewExitError("ethsign: missing required parameter --"+required, 1)
+					}
+				}
+
+				from := common.HexToAddress(c.String("from"))
+
+				dataString := c.String("data")
+				if !strings.HasPrefix(dataString, "0x") {
+					dataString = "0x" + dataString
+				}
+				data := hexutil.MustDecode(dataString)
+
+				wallets := getWallets(c.StringSlice("key-store"), defaultKeyStores)
+
+				var wallet accounts.Wallet
+				var acct *accounts.Account
+
+				needPassphrase := true
+
+			Scan:
```

Comment:
I suppose this loop and also the following passphrase stuff could also be refactored to a new function?

Reply:
Yeah, it's really the command inputs and eventual signing that changes between `tx` and `msg`. Everything in between is the same.